esp/esp/program/models/__init__.py

```diff
@@ -536,26 +536,121 @@ def volunteers_union(self, QObject = False):
+    def _student_is_in_program(self, user):
+        """Return whether the student is in the program."""
+        students = self.students()['classreg']
+        return students.filter(id=user.id).exists()
+    _student_is_in_program.depend_on_row('program.ClassSubject', lambda cls: {'self': cls.parent_program})
+    _student_is_in_program.depend_on_row('program.ClassSection', lambda cls: {'self': cls.parent_class.parent_program})
+    _student_is_in_program.depend_on_row('program.StudentRegistration', lambda sr: {'user': sr.user})
+    _student_is_in_program.depend_on_row('program.StudentSubjectInterest', lambda ssi: {'user': ssi.user})
+    _student_is_in_program.get_or_create_token(('self',))
+    _student_is_in_program.get_or_create_token(('user',))
+
+    def _user_can_join_by_grade(self, user):
+        """Helper function for user_can_join, when using program_size_by_grade.
+
+        This may be very slow if the user is not in the program; unfortunately
+        there's not much we can do about it.
+        """
+        # Check these first because computing the number of students in the
+        # program is slow and uncacheable.
+        if user.canRegToFullProgram(self):
+            return True
+        if self._student_is_in_program(user):
+            return True
+        caps = self._grade_caps()
+        grade = user.getGrade(self, assume_student=True)
+        for grades, cap in caps.iteritems():
+            if (grade in grades and
+                    self._students_in_program_in_grades(grades) >= cap):
+                return False
+        return True
+
     @cache_function
-    def isFull(self):
-        """ Can this program accept any more students? """
+    def _grade_caps(self):
+        """Parses the program_size_by_grade Tag.
 
-        # Some programs don't have caps; this is represented with program_size_max in [ 0, None ]
-        if self.program_size_max is None or self.program_size_max == 0:
-            return False
+        See user_can_join for the tag syntax.
 
-        students_dict = self.students(QObjects = True)
-        if students_dict.has_key('classreg'):
-            students_count = ESPUser.objects.filter(students_dict['classreg']).distinct().count()
-        else:
-            students_count = ESPUser.objects.filter(record__event="reg_confirmed",record__program=self).distinct().count()
+        Returns a dict with tuples of valid grades as keys, and caps as values.
+        """
+        size_tag = Tag.getProgramTag("program_size_by_grade", self)
+        size_dict = {}
+        for k, v in json.loads(size_tag).iteritems():
+            if '-' in k:
+                low, high = map(int, k.split('-'))
+                size_dict[tuple(xrange(low, high + 1))] = v
+            else:
+                size_dict[(int(k),)] = v
+        return size_dict
+    _grade_caps.depend_on_model('tagdict.Tag')
 
-        isfull = ( students_count >= self.program_size_max )
+    def _user_can_join_at_all(self, user):
```

Comment:
The caller has these lines:

```python
elif self.program_size_max:
    return self._user_can_join_at_all(user)
else:
    return True
```

But if we were to start calling ``_user_can_join_at_all`` from somewhere else, or if we refactored things, we could accidentally call ``_user_can_join_at_all()`` without doing the ``if self.program_size_max`` check. And this function would produce the incorrect result; it'd always return `False` (or raise an exception) if it reaches the last line.

Can we instead move the ``if self.program_size_max`` check out of the caller, and into the top of this function?